File: runtimes/ts/src/engine.ts
// SahouNode — zenoh-ts glue. All boundary semantics are delegated to the core (wasm) (design §1, option B).
// Symmetric with the Python version, runtimes/py/python/sahou/_engine.py.
import {
  CongestionControl,
  Duration,
  Priority,
  Query,
  Reliability,
  ReplyError,
  Sample,
  Session,
} from "@eclipse-zenoh/zenoh-ts";
import type { CoreModule, CoreRuntime } from "./core.js";
import { Diag, SahouRejected, SahouUnreachable, fmtDiags } from "./diag.js";

export type Json = Record<string, unknown>;
export type RejectHandler = (conn: string, diags: Diag[]) => void | Promise<void>;

export interface NodePlan {
  publishes: string[];
  subscribes: string[];
  queries: string[];
  answers: string[];
}

interface QosSpec {
  reliability: string;
  congestion: string;
  priority: string;
  express: boolean;
}

const PRIORITY: Record<string, Priority> = {
  realtime: Priority.REAL_TIME,
  interactive_high: Priority.INTERACTIVE_HIGH,
  interactive_low: Priority.INTERACTIVE_LOW,
  data_high: Priority.DATA_HIGH,
  data: Priority.DATA,
  data_low: Priority.DATA_LOW,
  background: Priority.BACKGROUND,
};

const utf8Strict = new TextDecoder("utf-8", { fatal: true });

/** Decode the wire attachment to a string. Non-UTF-8 (a non-sahou sender) yields undefined → the core returns missing_schema_hash. */
function decodeAttachment(att: { toBytes(): Uint8Array } | null | undefined): string | undefined {
  if (att == null) return undefined;
  try {
    return utf8Strict.decode(att.toBytes());
  } catch {
    return undefined;
  }
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

/** Convert a core throw (message = diags JSON) into a SahouRejected. Anything else is re-thrown as-is. */
export function toRejected(e: unknown): never {
  if (e instanceof Error) {
    try {
      throw new SahouRejected(JSON.parse(e.message) as Diag[]);
    } catch (parsed) {
      if (parsed instanceof SahouRejected) throw parsed;
    }
  }
  throw e;
}

export class SahouNode {
  readonly node: string;
  /** Diagnostic code → cumulative reject count (counted NO; never silently dropped). */
  readonly rejectCounts = new Map<string, number>();

  private readonly core: CoreModule;
  private readonly rt: CoreRuntime;
  private readonly session: Session;
  private readonly plan: NodePlan;
  private readonly ns: string;
  private readonly connInfo = new Map<string, { key: string; hash: string }>();
  private readonly publishers = new Map<string, { put(payload: string, opts: { attachment: string }): Promise<void> }>();
  private readonly txSeq = new Map<string, number>();
  private readonly rxSeq = new Map<string, number>();
  private readonly handles: { undeclare(): Promise<void> }[] = [];
  /** `${conn} ${hash}` -> the core's handshake verdict (a blocked verdict replays the core diags; Important-2). */
  private readonly verdicts = new Map<string, { verdict: "accepted" | "blocked"; diags: Diag[] }>();
  private readonly pending = new Set<string>();
  private onRejectGlobal: RejectHandler | undefined;

  static async create(core: CoreModule, session: Session, descriptorJson: string, node: string): Promise<SahouNode> {
    let rt: CoreRuntime;
    let plan: NodePlan;
    try {
      rt = new core.WasmRuntime(descriptorJson);
      plan = JSON.parse(rt.node_plan(node)) as NodePlan;
    } catch (e) {
      toRejected(e); // turn the core diagnostics (JSON) into an exception
    }
    const n = new SahouNode(core, rt, session, node, plan);
    await n.declareContracts();
    return n;
  }

  private constructor(core: CoreModule, rt: CoreRuntime, session: Session, node: string, plan: NodePlan) {
    this.core = core;
    this.rt = rt;
    this.session = session;
    this.node = node;
    this.plan = plan;
    this.ns = rt.namespace();
  }

  /** Contract queryable for every connection this node participates in (design §5-1: content-addressed; every participant declares it). */
  private async declareContracts(): Promise<void> {
    const conns = new Set([
      ...this.plan.publishes,
      ...this.plan.subscribes,
      ...this.plan.queries,
      ...this.plan.answers,
    ]);
    for (const conn of conns) {
      const fragJson = this.rt.contract_fragment(conn);
      const frag = JSON.parse(fragJson) as { key: string; hash: string };
      this.connInfo.set(conn, { key: frag.key, hash: frag.hash });
      const contractKey = `${this.ns}/@sahou/contract/${conn}/${frag.hash}`;
      const q = await this.session.declareQueryable(contractKey, {
        handler: async (query: Query) => {
          try {
            await query.reply(contractKey, fragJson);
          } catch (e) {
            console.warn(`[sahou] contract queryable reply failed on ${contractKey}`, e);
          } finally {
            await query.finalize();
          }
        },
      });
      this.handles.push(q);
    }
  }

  // ---- Public API -----------------------------------------------------

  connectionInfo(conn: string): { key: string; hash: string } {
    const info = this.connInfo.get(conn);
    if (!info) {
      throw new SahouRejected([
        { code: "role_mismatch", path: `connections.${conn}`, message: `node '${this.node}' does not participate in this connection` },
      ]);
    }
    return { ...info };
  }

  onReject(cb: RejectHandler): void {
    this.onRejectGlobal = cb;
  }

  async publish(conn: string, payload: Json): Promise<void> {
    const seq = this.nextSeq(this.txSeq, conn);
    const res = JSON.parse(this.rt.prepare_publish(this.node, conn, JSON.stringify(payload), BigInt(seq)));
    if (!res.ok) throw new SahouRejected(res.diags); // send-boundary NO → do not put
    const pub = await this.publisher(conn, res.msg.qos as QosSpec);
    await pub.put(res.msg.wire, { attachment: res.msg.attachment });
  }

  async subscribe(
    conn: string,
    handler: (payload: Json) => void | Promise<void>,
    opts: { onReject?: RejectHandler } = {},
  ): Promise<void> {
    if (!this.plan.subscribes.includes(conn)) {
      throw new SahouRejected([
        { code: "role_mismatch", path: `connections.${conn}`, message: `node '${this.node}' is not a receiver on this connection` },
      ]);
    }
    const key = this.connectionInfo(conn).key;
    const sub = await this.session.declareSubscriber(key, {
      handler: (sample: Sample) => this.handleSample(conn, handler, opts.onReject, sample),
    });
    this.handles.push(sub);
  }

  /** The requester side of query boundaries ①–④. Returns { delivered, response, diags, timedOut }. */
  async query(
    conn: string,
    payload: Json,
    opts: { timeoutMs?: number } = {},
  ): Promise<{ delivered: boolean; response: Json | null; diags: Diag[]; timedOut: boolean }> {
    const seq = this.nextSeq(this.txSeq, conn);
    const res = JSON.parse(this.rt.prepare_request(this.node, conn, JSON.stringify(payload), BigInt(seq)));
    if (!res.ok) throw new SahouRejected(res.diags); // ① send boundary = do not even issue the get
    let delivered = false;
    let response: Json | null = null;
    const diagsAll: Diag[] = [];
    let gotAny = false;
    try {
      const receiver = await this.session.get(res.msg.key, {
        payload: res.msg.wire,
        attachment: res.msg.attachment,
        timeout: Duration.milliseconds.of(opts.timeoutMs ?? 2000),
      });
      if (receiver) {
        for await (const reply of receiver) {
          gotAny = true;
          const r = reply.result();
          if (r instanceof ReplyError) {
            // envelope interpretation lives in a single core implementation (bad_reply_envelope = retryable)
            diagsAll.push(...(JSON.parse(this.core.wasm_parse_reply_err(r.payload().toBytes())) as Diag[]));
            continue;
          }
          const rseq = this.nextSeq(this.rxSeq, conn);
          const wire = r.payload().toBytes();
          const att = decodeAttachment(r.attachment?.());
          let out = JSON.parse(this.rt.accept_reply(this.node, conn, wire, att, BigInt(rseq), undefined));
          if (out.result === "hash_mismatch") out = this.resolveMismatch(conn, out.sender_hash, wire, rseq, "reply");
          if (out.result === "accept") {
            // ④ reply receive boundary
            delivered = true;
            response = JSON.parse(out.payload) as Json;
            break;
          }
          diagsAll.push(...(out.diags as Diag[]));
        }
      }
    } catch (e) {
      console.error(`[sahou] query get failed on '${conn}'`, e); // a failure of the get itself is treated like a timeout (retryable)
    }
    return { delivered, response, diags: diagsAll, timedOut: !gotAny };
  }

  /** Confirmed delivery (Z20): return on a 200-equivalent / fatal (4xx) throws SahouRejected immediately / retryable is resent. */
  async queryConfirmed(
    conn: string,
    payload: Json,
    opts: { timeoutMs?: number; retries?: number; backoffMs?: number } = {},
  ): Promise<Json> {
    const retries = opts.retries ?? 3;
    const backoff = opts.backoffMs ?? 300;
    for (let attempt = 1; ; attempt++) {
      const r = await this.query(conn, payload, { timeoutMs: opts.timeoutMs });
      if (r.delivered) return r.response as Json;
      const cls = this.core.wasm_classify_delivery(r.timedOut, JSON.stringify(r.diags));
      if (cls === "fatal") throw new SahouRejected(r.diags);
      if (attempt > retries) throw new SahouUnreachable(conn, attempt);
      await sleep(backoff * attempt);
    }
  }

  /** The responder side of query (② receive boundary + ③ reply send boundary). */
  async answer(conn: string, fn: (req: Json) => Json | Promise<Json>): Promise<void> {
    if (!this.plan.answers.includes(conn)) {
      throw new SahouRejected([
        { code: "role_mismatch", path: `connections.${conn}`, message: `node '${this.node}' is not the responder on this connection` },
      ]);
    }
    const key = this.connectionInfo(conn).key;
    const q = await this.session.declareQueryable(key, {
      handler: async (query: Query) => {
        try {
          const rseq = this.nextSeq(this.rxSeq, conn);
          const p = query.payload();
          const wire = p ? p.toBytes() : new Uint8Array();
          const att = decodeAttachment(query.attachment?.());
          let out = JSON.parse(this.rt.accept_request(this.node, conn, wire, att, BigInt(rseq), undefined));
          if (out.result === "hash_mismatch") out = this.resolveMismatch(conn, out.sender_hash, wire, rseq, "request");
          if (out.result !== "accept") {
            // ② request receive boundary
            await query.replyErr(JSON.stringify({ diags: out.diags }));
            return;
          }
          let resp: Json;
          try {
            resp = await fn(JSON.parse(out.payload) as Json);
          } catch (e) {
            // a handler exception is returned as a 5xx-equivalent (retryable)
            await query.replyErr(JSON.stringify({ diags: [{ code: "handler_error", path: "$", message: String(e) }] }));
            return;
          }
          const tseq = this.nextSeq(this.txSeq, conn);
          const res = JSON.parse(this.rt.prepare_reply(this.node, conn, JSON.stringify(resp), BigInt(tseq)));
          if (!res.ok) {
            // ③ reply send boundary = do not reply with a broken response
            await query.replyErr(JSON.stringify({ diags: res.diags }));
            return;
          }
          await query.reply(res.msg.key, res.msg.wire, { attachment: res.msg.attachment });
        } catch (e) {
          console.error(`[sahou] internal error in queryable '${conn}'`, e); // last line of defense
        } finally {
          await query.finalize(); // close the querier's channel (013-3: required for Node)
        }
      },
    });
    this.handles.push(q);
  }

  async close(): Promise<void> {
    for (const h of this.handles) {
      try {
        await h.undeclare();
      } catch (e) {
        console.warn("[sahou] undeclare failed", e); // close is best-effort (log, don't swallow)
      }
    }
    try {
      await this.session.close();
    } catch (e) {
      console.warn("[sahou] session close failed", e);
    }
    // close is terminal (rt is not used afterwards). Free the core instance held for the whole lifetime here too.
    try {
      this.rt.free();
    } catch (e) {
      console.warn("[sahou] rt.free failed", e); // best-effort (log, don't swallow)
    }
  }

  // ---- Internals ------------------------------------------------------

  private nextSeq(counter: Map<string, number>, conn: string): number {
    const seq = counter.get(conn) ?? 0;
    counter.set(conn, seq + 1);
    return seq;
  }

  private async publisher(conn: string, qos: QosSpec) {
    let pub = this.publishers.get(conn);
    if (!pub) {
      const key = this.connectionInfo(conn).key;
      const prio = PRIORITY[qos.priority];
      if (prio === undefined) {
        // the descriptor enum is guaranteed by the core, so reaching here means a missing mapping in the glue (do not silently fall back)
        throw new SahouRejected([
          { code: "descriptor_error", path: "$", message: `unknown priority '${qos.priority}' (missing QoS mapping in the glue)` },
        ]);
      }
      pub = await this.session.declarePublisher(key, {
        reliability: qos.reliability === "reliable" ? Reliability.RELIABLE : Reliability.BEST_EFFORT,
        congestionControl: qos.congestion === "block" ? CongestionControl.BLOCK : CongestionControl.DROP,
        priority: prio,
        express: qos.express,
      });
      this.publishers.set(conn, pub);
    }
    return pub;
  }

  private countReject(conn: string, diags: Diag[], onReject?: RejectHandler): void {
    for (const d of diags) this.rejectCounts.set(d.code, (this.rejectCounts.get(d.code) ?? 0) + 1);
    const cb = onReject ?? this.onRejectGlobal;
    if (cb) {
      try {
        // sync throws are caught here; an async callback's rejection is caught by .catch() (both log only; the receive path continues)
        Promise.resolve(cb(conn, diags)).catch((e) => {
          console.error("[sahou] onReject callback failed", e);
        });
      } catch (e) {
        console.error("[sahou] onReject callback failed", e); // a failing user callback must not kill the receive path
      }
    } else {
      console.warn(`[sahou] reject on '${conn}': ${fmtDiags(diags)}`);
    }
  }

  private handleSample(
    conn: string,
    handler: (payload: Json) => void | Promise<void>,
    onReject: RejectHandler | undefined,
    sample: { payload(): { toBytes(): Uint8Array }; attachment?(): { toBytes(): Uint8Array } | undefined },
  ): void {
    try {
      const seq = this.nextSeq(this.rxSeq, conn);
      const wire = sample.payload().toBytes();
      const att = decodeAttachment(sample.attachment?.());
      let out = JSON.parse(this.rt.accept_sample(this.node, conn, wire, att, BigInt(seq), undefined));
      if (out.result === "hash_mismatch") out = this.resolveMismatch(conn, out.sender_hash, wire, seq, "sample");
      if (out.result === "accept") {
        try {
          // sync throws are caught here; an async handler's rejection is caught by .catch() (both log only; the receive path continues)
          Promise.resolve(handler(JSON.parse(out.payload) as Json)).catch((e) => {
            console.error(`[sahou] handler failed on '${conn}'`, e);
          });
        } catch (e) {
          console.error(`[sahou] handler failed on '${conn}'`, e); // a handler exception must not kill the receive path
        }
      } else {
        this.countReject(conn, out.diags as Diag[], onReject);
      }
    } catch (e) {
      console.error(`[sahou] internal error while handling sample on '${conn}'`, e); // last line of defense (don't die silently)
    }
  }

  private resolveMismatch(
    conn: string,
    senderHash: string,
    wire: Uint8Array,
    seq: number,
    kind: "sample" | "request" | "reply",
  ): { result: string; diags?: Diag[]; payload?: string } {
    const k = `${conn} ${senderHash}`;
    const entry = this.verdicts.get(k);
    if (entry) {
      if (entry.verdict === "accepted") {
        const accept = {
          sample: this.rt.accept_sample,
          request: this.rt.accept_request,
          reply: this.rt.accept_reply,
        }[kind].bind(this.rt);
        return JSON.parse(accept(this.node, conn, wire, senderHash, BigInt(seq), senderHash));
      }
      // blocked: replay the core's handshake diags verbatim (no glue-authored wording; byte-identical across all three languages)
      return { result: "reject", diags: entry.diags };
    }
    void this.startHandshake(conn, senderHash).catch((e) => {
      console.error(`[sahou] handshake failed on '${conn}'`, e);
    });
    return {
      result: "reject",
      diags: [
        {
          code: "handshake_pending",
          path: `connections.${conn}`,
          message: `contract version mismatch detected; handshake in progress (sender_hash=${senderHash})`,
        },
      ],
    };
  }

  /** Design §5-2 to 5-4: fetch the contract once and cache the core's three-valued verdict. `unreachable` is not cached. */
  private async startHandshake(conn: string, senderHash: string): Promise<void> {
    const k = `${conn} ${senderHash}`;
    if (this.pending.has(k) || this.verdicts.has(k)) return;
    this.pending.add(k);
    try {
      const sel = `${this.ns}/@sahou/contract/${conn}/${senderHash}`;
      let fragment: string | null = null;
      try {
        const receiver = await this.session.get(sel, { timeout: Duration.milliseconds.of(2000) });
        if (receiver) {
          for await (const reply of receiver) {
            const r = reply.result();
            if (!(r instanceof ReplyError)) {
              fragment = r.payload().toString();
              break;
            }
          }
        }
      } catch (e) {
        console.warn(`[sahou] contract fetch failed: ${sel}`, e); // a failure of the get itself is treated as unreachable
      }
      if (fragment == null) {
        console.warn(`[sahou] [contract_unreachable] ${sel}: cannot fetch the contract (not cached; will retry on the next mismatch detection)`);
        return;
      }
      const res = JSON.parse(this.rt.handshake(conn, senderHash, fragment));
      if (res.verdict === "unreachable") {
        console.warn(`[sahou] [contract_unreachable] '${conn}' (sender=${senderHash}): ${fmtDiags(res.diags)} (not cached)`);
        return;
      }
      this.verdicts.set(k, { verdict: res.verdict, diags: (res.diags ?? []) as Diag[] });
      if (res.verdict === "accepted") {
        console.info(`[sahou] handshake accepted on '${conn}' (sender=${senderHash}, additive)`);
      } else {
        console.warn(`[sahou] [schema_incompatible] '${conn}' (sender=${senderHash}): ${fmtDiags(res.diags)}`);
      }
    } finally {
      this.pending.delete(k);
    }
  }
}

File: runtimes/python/python/sahou/_engine.py
"""SahouNode — zenoh glue. All boundary semantics are delegated to the core (sahou._core) (design §1, option B)."""
from __future__ import annotations

import json
import logging
import threading
import time
from collections import Counter
from pathlib import Path

import zenoh

from ._core import SahouRuntime, classify_delivery, parse_reply_err
from ._diag import SahouRejected, SahouUnreachable

log = logging.getLogger("sahou")

_PRIORITY = {
    "realtime": zenoh.Priority.REAL_TIME,
    "interactive_high": zenoh.Priority.INTERACTIVE_HIGH,
    "interactive_low": zenoh.Priority.INTERACTIVE_LOW,
    "data_high": zenoh.Priority.DATA_HIGH,
    "data": zenoh.Priority.DATA,
    "data_low": zenoh.Priority.DATA_LOW,
    "background": zenoh.Priority.BACKGROUND,
}


def _read_descriptor(descriptor) -> str:
    if isinstance(descriptor, dict):
        return json.dumps(descriptor)
    text = str(descriptor)
    p = Path(text)
    if p.suffix == ".json" and p.exists():
        return p.read_text(encoding="utf-8")
    return text  # the JSON string itself


def _fmt_diags(diags: list[dict]) -> str:
    return "; ".join(f"[{d['code']}] @{d['path']}: {d['message']}" for d in diags)


def _decode_attachment(att_z):
    """Decode a wire attachment to str. Non-UTF-8 (a non-sahou sender) becomes None → the core returns missing_schema_hash."""
    if att_z is None:
        return None
    try:
        return bytes(att_z).decode()
    except UnicodeDecodeError:
        return None


def connect(descriptor, node: str, *, connect=None, listen=None, multicast: bool = True) -> "SahouNode":
    """Connect with a descriptor (path / JSON string / dict) plus a node name. Default is automatic LAN discovery (multicast peer)."""
    desc_json = _read_descriptor(descriptor)
    try:
        rt = SahouRuntime(desc_json)
        plan = json.loads(rt.node_plan(node))
    except ValueError as e:  # turn the core's diagnostics (JSON) into an exception
        raise SahouRejected(json.loads(str(e))) from None
    conf = zenoh.Config()
    if connect:
        conf.insert_json5("connect/endpoints", json.dumps(list(connect)))
    if listen:
        conf.insert_json5("listen/endpoints", json.dumps(list(listen)))
    if not multicast:
        conf.insert_json5("scouting/multicast/enabled", "false")
    session = zenoh.open(conf)
    listen_port = None
    # Test aid: remember the rendezvous port of the mesh this node joins
    # (if listening, that port; if only connecting, the target port — i.e. the only entry point
    #  through which a raw session can join the same mesh and have a route to this node).
    endpoints = listen or connect
    if endpoints:
        try:
            listen_port = int(str(endpoints[0]).rsplit(":", 1)[1])
        except (IndexError, ValueError):
            listen_port = None
    return SahouNode(rt, node, session, plan, listen_port)


class SahouNode:
    def __init__(self, rt: SahouRuntime, node: str, session, plan: dict, listen_port=None):
        self._rt = rt
        self._node = node
        self._session = session
        self._plan = plan
        self._ns = rt.namespace()
        self._listen_port = listen_port  # test aid (for raw bypass)
        self._publishers: dict[str, object] = {}
        self._tx_seq: Counter = Counter()
        self._rx_seq: Counter = Counter()
        self._handles: list = []  # keep subscribers/queryables alive
        self._conn_info: dict[str, dict] = {}
        self._verdicts: dict[tuple[str, str], tuple[str, list]] = {}  # (conn, sender_hash) -> (verdict, core diags)
        self._pending: set[tuple[str, str]] = set()
        self._lock = threading.Lock()
        self.reject_counts: Counter = Counter()
        self._on_reject_global = None
        # contract queryable for every connection this node joins (design §5-1: content-addressed, declared by all participants)
        for conn in set(plan["publishes"]) | set(plan["subscribes"]) | set(plan["queries"]) | set(plan["answers"]):
            frag_json = rt.contract_fragment(conn)
            frag = json.loads(frag_json)
            self._conn_info[conn] = {"key": frag["key"], "hash": frag["hash"]}
            contract_key = f"{self._ns}/@sahou/contract/{conn}/{frag['hash']}"
            self._handles.append(
                self._session.declare_queryable(contract_key, self._make_contract_cb(contract_key, frag_json))
            )

    # ---- public API -----------------------------------------------------

    def connection_info(self, conn: str) -> dict:
        return dict(self._conn_info[conn])

    def on_reject(self, cb):
        self._on_reject_global = cb

    def publish(self, conn: str, payload: dict) -> None:
        seq = self._next_seq(self._tx_seq, conn)
        res = json.loads(self._rt.prepare_publish(self._node, conn, json.dumps(payload), seq))
        if not res["ok"]:
            raise SahouRejected(res["diags"])
        msg = res["msg"]
        self._publisher(conn, msg["qos"]).put(msg["wire"].encode(), attachment=msg["attachment"].encode())

    def subscribe(self, conn: str, handler=None, *, on_reject=None):
        def register(fn):
            key = self._conn_info[conn]["key"]  # a connection not joined raises KeyError → made an explicit error
            def cb(sample):
                self._handle_sample(conn, fn, on_reject, sample)
            sub = self._session.declare_subscriber(key, cb)
            self._handles.append(sub)
            return fn
        if conn not in self._conn_info or conn not in self._plan["subscribes"]:
            raise SahouRejected([{"code": "role_mismatch", "path": f"connections.{conn}",
                                  "message": f"node '{self._node}' is not a receiver on this connection"}])
        return register(handler) if handler is not None else register

    def query(self, conn: str, payload: dict, timeout: float = 2.0) -> dict:
        """Requester side of the query ①-④ boundaries. Returns {delivered, response, diags, timed_out}."""
        seq = self._next_seq(self._tx_seq, conn)
        res = json.loads(self._rt.prepare_request(self._node, conn, json.dumps(payload), seq))
        if not res["ok"]:
            raise SahouRejected(res["diags"])  # ① send boundary = do not even issue the get
        msg = res["msg"]
        delivered, response, diags_all, got_any = False, None, [], False
        try:
            replies = self._session.get(
                msg["key"], payload=msg["wire"].encode(), attachment=msg["attachment"].encode(), timeout=timeout
            )
            for reply in replies:
                got_any = True
                sample = getattr(reply, "ok", None)
                if sample is not None:
                    rseq = self._next_seq(self._rx_seq, conn)
                    wire = bytes(sample.payload)
                    att_z = sample.attachment
                    att = _decode_attachment(att_z)
                    out = json.loads(self._rt.accept_reply(self._node, conn, wire, att, rseq, None))
                    if out["result"] == "hash_mismatch":
                        out = self._resolve_mismatch(conn, att, wire, rseq, kind="reply")
                    if out["result"] == "accept":  # ④ reply-receive boundary
                        delivered, response = True, json.loads(out["payload"])
                        break
                    diags_all.extend(out["diags"])
                else:
                    err = reply.err
                    diags_all.extend(json.loads(parse_reply_err(bytes(err.payload))))
        except Exception:  # noqa: BLE001 - a failure of the get itself is treated as a timeout (retryable)
            log.exception("query get failed on '%s'", conn)
        return {"delivered": delivered, "response": response, "diags": diags_all,
                "timed_out": not got_any}

    def query_confirmed(self, conn: str, payload: dict, *, timeout: float = 2.0,
                        retries: int = 3, backoff: float = 0.3) -> dict:
        """Confirmed delivery (Z20): return on a 200 equivalent / raise SahouRejected immediately on fatal (4xx) / resend on retryable."""
        attempt = 0
        while True:
            attempt += 1
            r = self.query(conn, payload, timeout=timeout)
            if r["delivered"]:
                return r["response"]
            cls = classify_delivery(r["timed_out"], json.dumps(r["diags"]))
            if cls == "fatal":
                raise SahouRejected(r["diags"])
            if attempt > retries:
                raise SahouUnreachable(conn, attempt)
            time.sleep(backoff * attempt)

    def answer(self, conn: str, fn):
        """Responder side of a query (② receive boundary + ③ reply-send boundary). Usable as a decorator too."""
        if conn not in self._plan["answers"]:
            raise SahouRejected([{"code": "role_mismatch", "path": f"connections.{conn}",
                                  "message": f"node '{self._node}' is not a responder on this connection"}])
        key = self._conn_info[conn]["key"]

        def qcb(query):
            try:
                rseq = self._next_seq(self._rx_seq, conn)
                payload_z = query.payload
                wire = bytes(payload_z) if payload_z is not None else b""
                att_z = query.attachment
                att = _decode_attachment(att_z)
                out = json.loads(self._rt.accept_request(self._node, conn, wire, att, rseq, None))
                if out["result"] == "hash_mismatch":
                    out = self._resolve_mismatch(conn, att, wire, rseq, kind="request")
                if out["result"] != "accept":  # ② request-receive boundary
                    query.reply_err(json.dumps({"diags": out["diags"]}).encode())
                    return
                try:
                    resp = fn(json.loads(out["payload"]))
                except Exception as e:  # noqa: BLE001 - a handler exception is returned as a 5xx equivalent
                    query.reply_err(json.dumps({"diags": [{"code": "handler_error", "path": "$",
                                                           "message": str(e)}]}).encode())
                    return
                tseq = self._next_seq(self._tx_seq, conn)
                res = json.loads(self._rt.prepare_reply(self._node, conn, json.dumps(resp), tseq))
                if not res["ok"]:  # ③ reply-send boundary = do not reply with a broken response
                    query.reply_err(json.dumps({"diags": res["diags"]}).encode())
                    return
                msg = res["msg"]
                query.reply(msg["key"], msg["wire"].encode(), attachment=msg["attachment"].encode())
            except Exception:  # noqa: BLE001 - last line of defense
                log.exception("internal error in queryable '%s'", conn)

        self._handles.append(self._session.declare_queryable(key, qcb))
        return fn

    def close(self):
        for h in self._handles:
            try:
                h.undeclare()
            except Exception:  # noqa: BLE001 - close is best-effort (diagnostics are logged)
                log.debug("undeclare failed", exc_info=True)
        try:
            self._session.close()
        except Exception:  # noqa: BLE001 - session.close occasionally times out during loopback teardown.
            # Best-effort since this is teardown (the OS reclaims the socket). Log it rather than swallow it.
            log.debug("session close failed", exc_info=True)

    # ---- internals ------------------------------------------------------

    def _next_seq(self, counter, conn: str) -> int:
        """Allocate the per-connection sequence number atomically (safe across zenoh callbacks / multiple threads).
        The query/answer seq of Tasks 8/9 also use this helper."""
        with self._lock:
            seq = counter[conn]
            counter[conn] += 1
            return seq

    def _publisher(self, conn: str, qos: dict):
        if conn not in self._publishers:
            key = self._conn_info[conn]["key"]
            self._publishers[conn] = self._session.declare_publisher(
                key,
                reliability=zenoh.Reliability.RELIABLE if qos["reliability"] == "reliable" else zenoh.Reliability.BEST_EFFORT,
                congestion_control=zenoh.CongestionControl.BLOCK if qos["congestion"] == "block" else zenoh.CongestionControl.DROP,
                priority=_PRIORITY[qos["priority"]],
                express=qos["express"],
            )
        return self._publishers[conn]

    def _count_reject(self, conn: str, diags: list[dict], on_reject):
        with self._lock:
            for d in diags:
                self.reject_counts[d["code"]] += 1
        cb = on_reject or self._on_reject_global
        if cb is not None:
            try:
                cb(conn, diags)
            except Exception:  # noqa: BLE001 - a failure in the user callback must not kill the zenoh thread
                log.exception("on_reject callback failed")
        else:
            log.warning("sahou reject on '%s': %s", conn, _fmt_diags(diags))

    def _handle_sample(self, conn: str, fn, on_reject, sample):
        try:
            seq = self._next_seq(self._rx_seq, conn)
            wire = bytes(sample.payload)
            att_z = sample.attachment
            att = _decode_attachment(att_z)
            out = json.loads(self._rt.accept_sample(self._node, conn, wire, att, seq, None))
            if out["result"] == "hash_mismatch":
                out = self._resolve_mismatch(conn, att, wire, seq, kind="sample")
            if out["result"] == "accept":
                try:
                    fn(json.loads(out["payload"]))
                except Exception:  # noqa: BLE001 - a handler exception must not kill the zenoh thread
                    log.exception("handler failed on '%s'", conn)
            else:
                self._count_reject(conn, out["diags"], on_reject)
        except Exception:  # noqa: BLE001 - last line of defense (do not die silently)
            log.exception("internal error while handling sample on '%s'", conn)

    def _resolve_mismatch(self, conn: str, sender_hash: str, wire: bytes, seq: int, *, kind: str) -> dict:
        """Look up the verdict cache; if unknown, start a handshake and return a handshake_pending reject."""
        with self._lock:
            entry = self._verdicts.get((conn, sender_hash))
        if entry is not None:
            verdict, diags = entry
            if verdict == "accepted":
                accept = {
                    "sample": self._rt.accept_sample,
                    "request": self._rt.accept_request,
                    "reply": self._rt.accept_reply,
                }[kind]
                return json.loads(accept(self._node, conn, wire, sender_hash, seq, sender_hash))
            # blocked: replay the core's handshake-verdict diags as-is (no glue-authored wording; byte-identical across the 3 languages)
            return {"result": "reject", "diags": diags}
        self._start_handshake(conn, sender_hash)
        return {"result": "reject", "diags": [{"code": "handshake_pending", "path": f"connections.{conn}",
                                               "message": f"contract version mismatch detected; handshake in progress (sender_hash={sender_hash})"}]}

    def _make_contract_cb(self, contract_key: str, fragment_json: str):
        def cb(query):
            try:
                query.reply(contract_key, fragment_json.encode())
            except Exception:  # noqa: BLE001 - a failure to reply to the queryable must not kill the zenoh thread
                log.exception("contract queryable reply failed on %s", contract_key)
        return cb

    def _start_handshake(self, conn: str, sender_hash: str) -> None:
        key = (conn, sender_hash)
        with self._lock:
            if key in self._pending or key in self._verdicts:
                return
            self._pending.add(key)
        threading.Thread(target=self._fetch_and_judge, args=(conn, sender_hash), daemon=True).start()

    def _fetch_and_judge(self, conn: str, sender_hash: str) -> None:
        """Design §5-2 to 5-4: fetch the peer's contract exactly once and cache the verdict from the core's 3-way decision.
        unreachable (unfetchable / undecodable / fragment hash mismatch) is not cached = retried on the next mismatch detection."""
        try:
            sel = f"{self._ns}/@sahou/contract/{conn}/{sender_hash}"
            fragment = None
            try:
                for reply in self._session.get(sel, timeout=2.0):
                    sample = getattr(reply, "ok", None)
                    if sample is not None:
                        fragment = bytes(sample.payload).decode()
                        break
            except Exception:  # noqa: BLE001 - a failure of the get itself is treated as unreachable
                log.exception("contract fetch failed: %s", sel)
            if fragment is None:
                log.warning("[contract_unreachable] %s: cannot fetch the peer's contract (not cached; retried on the next mismatch detection)", sel)
                return
            res = json.loads(self._rt.handshake(conn, sender_hash, fragment))
            if res["verdict"] == "unreachable":
                log.warning("[contract_unreachable] '%s' (sender=%s): %s (not cached; retried on the next mismatch detection)",
                            conn, sender_hash, _fmt_diags(res["diags"]))
                return
            with self._lock:
                self._verdicts[(conn, sender_hash)] = (res["verdict"], res.get("diags", []))
            if res["verdict"] == "accepted":
                log.info("handshake accepted on '%s' (sender=%s, additive)", conn, sender_hash)
            else:
                log.warning("[schema_incompatible] '%s' (sender=%s): %s", conn, sender_hash, _fmt_diags(res.get("diags", [])))
        except Exception:  # noqa: BLE001 - an unexpected exception is also treated as unreachable (not cached; do not die silently)
            log.exception(
                "[contract_unreachable] '%s' (sender=%s): unexpected exception during handshake (not cached; retried on the next mismatch detection)",
                conn, sender_hash,
            )
        finally:
            with self._lock:
                self._pending.discard((conn, sender_hash))
